anomalib/data/base.py

```diff
@@ -107,92 +176,34 @@ def __init__(
-        the path the ground truth masks (for the anomalous images only).
+        self.data: Optional[AnomalibDataset] = None
 
-        Example of a dataframe returned by calling this method from a concrete class:
-        |---|-------------------|-----------|-------------|------------------|-------|
-        |   | image_path        | label     | label_index | mask_path        | split |
-        |---|-------------------|-----------|-------------|------------------|-------|
-        | 0 | path/to/image.png | anomalous | 0           | path/to/mask.png | train |
-        |---|-------------------|-----------|-------------|------------------|-------|
-        """
+    @abstractmethod
+    def create_dataset(self) -> AnomalibDataset:
         raise NotImplementedError
 
-    def get_samples(self, split: Optional[str] = None) -> DataFrame:
-        """Retrieve the samples of the full dataset or one of the splits (train, val, test).
+    def prepare_data(self) -> None:
+        self.data = self.create_dataset()
 
-        Args:
-            split: (str): The split for which we want to retrieve the samples ("train", "val" or "test"). When
-                left empty, all samples will be returned.
-
-        Returns:
-            DataFrame: A dataframe containing the samples of the split or full dataset.
-        """
-        assert self._samples is not None, "Samples have not been created yet."
-        if split is None:
-            return self._samples
-        samples = self._samples[self._samples.split == split]
-        return samples.reset_index(drop=True)
+    def contains_anomalous_images(self, split):
+        samples = self.data.get_samples(split)
+        return 1 in list(samples.label_index)
 
-    def setup(self, stage: Optional[str] = None) -> None:
+    def setup(self, stage: Optional[str] = None):
         """Setup train, validation and test data.
 
         Args:
           stage: Optional[str]:  Train/Val/Test stages. (Default value = None)
         """
-        self._samples = self._create_samples()
-
-        logger.info("Setting up train, validation, test and prediction datasets.")
         if stage in (None, "fit"):
-            samples = self.get_samples("train")
-            self.train_data = AnomalibDataset(
-                samples=samples,
-                split="train",
-                task=self.task,
-                pre_process=self.pre_process_train,
-            )
-
+            self.train_data = self.data.train_subset(pre_process=self.pre_process_train)
         if stage in (None, "fit", "validate"):
-            samples = self.get_samples("val") if self.create_validation_set else self.get_samples("test")
-            self.val_data = AnomalibDataset(
-                samples=samples,
-                split="val",
-                task=self.task,
-                pre_process=self.pre_process_val,
-            )
-
+            if self.contains_anomalous_images("val"):
```

Comment:
Maybe we should use `subset.val` here as this calls `get_samples` which has subset as it's parameter